Fixed #20347 -- Allowed customizing the maximum number of instantiated forms in formsets.

smithdc1 opened the pull request:
[Ticket #20347](https://code.djangoproject.com/ticket/20347)

This follows on from #1593, I've updated the pull request based on the comments at the time and added release notes. 

Hopefully I've also added the original author as co-author correctly 🤞

felixxm commented on docs/topics/forms/formsets.txt:
```diff
@@ -126,6 +126,46 @@ affect validation.  If ``validate_max=True`` is passed to the
+
+    >>> from django.forms.formsets import formset_factory
+    >>> from myapp.forms import ArticleForm
+    >>> ArticleFormSet = formset_factory(ArticleForm, absolute_max=10)
```
This raises ``ValueError('absolute_max must be..`` because `DEFAULT_MAX_NUM = 1000`.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -81,6 +81,13 @@ def prefixed(*args):
+    # the formset has the proper prefix
+    def validate_formset_number_of_forms(self, formset, number_of_forms, prefix='form'):
+        self.assertEqual(number_of_forms, len(formset.forms))
+        for index in range(number_of_forms):
+            self.assertEquals('{0}-{1}'.format(prefix, index), formset.forms[index].prefix)
```
I'm not sure if this assertions have value.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+        # 2*DEFAULT_MAX_NUM if max_num isn't specified)
+        formset = FavoriteDrinksFormSet(data=data)
+        self.assertFalse(formset.is_valid())
+        self.validate_formset_number_of_forms(formset, 2 * formsets.DEFAULT_MAX_NUM)
```
This test works without a patch, please move it to a separate commit/PR, we should also add assertion for `absolute_max`, e.g.
```python
    def test_default_absolute_max(self):
        # absolute_max defaults to 2 * DEFAULT_MAX_NUM if max_num is None.
        data = {
            'form-TOTAL_FORMS': 2001,
            'form-INITIAL_FORMS': '0',
            'form-MAX_NUM_FORMS': '0',
        }
        formset = FavoriteDrinksFormSet(data=data)
        self.assertIs(formset.is_valid(), False)
        self.assertEqual(
            formset.non_form_errors(),
            ['Please submit 1000 or fewer forms.'],
        )
        self.assertEqual(formset.absolute_max, 2000)
```

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+        """
+        absolute_max can now be explicitly specified via kwarg to formset_factory.
+        This behavior was changed from absolute_max always being max_num + 1000 in the patch for #20347
+        """
```
Chop this docstring.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+        }
+
+        # absolute_max can now be explicitly set, not being affected by
+        # max_num or DEFAULT_MAX_NUM at all
```
I don't see any value in this comment.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+            'form-INITIAL_FORMS': '0',
+            'form-MAX_NUM_FORMS': '0',
+        }
+
```
Please chop unnecessary blank lines.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+        self.assertTrue(formset.is_valid())
+        self.validate_formset_number_of_forms(formset, 2505)
+
+        # verify that absolute_max still does provide a hard limit
```
Please omit prefixes like "Verify that" since all tests verify things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+        self.validate_formset_number_of_forms(formset, 30)
+
+    def test_explicit_absolute_max_lower_than_default_no_max_num(self):
+        with self.assertRaises(ValueError):
```
Use `assertRaisesMessage()`.

felixxm commented on tests/forms_tests/tests/test_formsets.py:
```diff
@@ -877,6 +884,67 @@ def test_more_initial_than_max_num(self):
+
+    def test_max_num_greater_than_absolute_max(self):
+        with self.assertRaises(ValueError):
+            formset_factory(FavoriteDrinkForm, max_num=100, absolute_max=30)
```
Use `assertRaisesMessage()`.

smithdc1 commented:
Hi @felixxm - Thanks for your feedback 🙏 

I've updated for your comments. I left the original commit alone and added my changes to subsequent commits. I was a bit unsure on rebasing for the release notes (as I had to force push?)

I've added some tests for ``modelformset_factory()`` and ``inlineformset_factory()`` but didn't think we need to test every permutation again?

felixxm commented on docs/topics/forms/formsets.txt:
```diff
@@ -126,6 +126,38 @@ affect validation.  If ``validate_max=True`` is passed to the
 
+.. _formsets-absolute-max:
+
+Limiting the maximum number of instantiated forms
```
I've changed `submitted` to `instantiated`.
